In this notebook we introduce basics of neural networks.

### Basics of ANN

Neural networks are a set of algorithms, modeled loosely after the human brain, that are designed to recognize patterns. Neural networks are made of perceptrons.

Perceptrons are cell units that are responsible for some mathematical computation.

<img src="https://miro.medium.com/max/1400/1*ZoT34FHnrrPxWHsXv6mK8Q.jpeg" width="60%">

In Artificial neural network, we use multiple such perceptrons to learn different attributes of the dataset. Together, they can learn highly complex non linear structure from data.

In the above picture - 

(1) is the input features

(2) are the weights of the hidden layer perceptrons

(3) are the representations learned by each perceptron

(4) is the activation function

(5) is the output


### Neural network architecture

Different types of cells are meant for different types of computations. e.g. - 

* Dense cell - these are meant for dense computations (basic perceptron)
* RNN cell - recurrent neural network cell is used where we have to store some temporal information in the NN
* Convolution cell - convolution cells are just like image filters, which are used to learn new features. Conv cells are mostly popular for image type of data.

Each cell takes input of a fixed dimension and generates intermediate hidden output of fixed dimension. During building the neural network, we need to make sure of the cell inputs and outputs, so that the final layer output matches with the data output dimensions.

In deep neural networks, multiple layer of multiple such cells are stacked together. e.g. - in the above diagram, we have a simple 1-layer ANN with just dense cells.

During the learning process, neural networks learn the weights of each of these cells. These are called the parameters of the neural network.

### Activation function

Activation functions are mathematical functions that are used in the hidden layers and in the final layer. By tuning activation functions, we learn more complex and non linear structures from data. Few activation functions are - 

* Linear
* Sigmoid (logistic)
* ReLU (rectified linear unit)
* Softmax
* Tanh
* Swish
* Mish
* ELU

<img src="https://miro.medium.com/max/2000/1*4ZEDRpFuCIpUjNgjDdT2Lg.png" width="50%">

ReLU is widely used activation function for intermediate layers. Sigmoid is widely used in the final layer for binary classification task. For multiclass classification, softmax is used in the final layer. In regression tasks, mostly ReLU is used in the final layer.

### How does neural network learn

In neural networks, we also need to define a loss function that calculates the loss between the actual prediction and the predicted output by the neural network. Once, loss is calculated, neural network uses $\textbf{back propagation}$ to learn the weights of the cells in it.

Forward propagation is the process of moving forward through the neural network (from inputs to the ultimate output or prediction). Backpropagation is the reverse. Except instead of signal, we are moving error backwards through our model.
 
<img src=https://miro.medium.com/max/1400/1*UY4-RIrSVgfuhAkawKIr2w.jpeg>

<img src=https://miro.medium.com/max/1400/1*0RIBu3Iz-aOOX9dyob_FHA.jpeg>

Back propagation can be thought as a kind of feedback loop for the network.

### Cost function

Cost function is used to calculate loss between actual and predicted output by the network.

<b> P.S. - although cost function and evaluation metric are very similar, there is a fine difference between two. cost function is used for backpropagation, due to which, it needs to differentiable. On the other hand, evaluation metric is used to evaluate a model's performance. </b> 

For regression the most popular loss function is Mean Squared Error (MSE) or L2 loss.

<img src="https://miro.medium.com/max/1026/1*SGhoeJ_BgcfqU06CmX41rw.png" width="20%">

Similarly, L1 loss or, Mean Absolute Error (MAE) is also used for regression task.

<img src="https://miro.medium.com/max/1066/1*piCo0iDgPmESnQkHSwAK6A.png" width="20%">

For classification, cross entropy loss is used.

<img src="https://miro.medium.com/max/1400/1*zi1wKAAGGt1Bn6mqo2MSFw.png" width="40%">

Other loss function are -

* KL divergence
* Hinge loss
* Triplet loss

### Optimization

Now that we have the loss calculated, how does NNs learn the model parameters?

As we need to learn model parameters that minimizes the overall loss, we treat the learning process as an optimization technique which updates the parameters in order to find the local (global) minima.

The name for one commonly used optimization function that adjusts weights according to the error they caused is called <b> gradient descent </b>.

Gradient is another word for slope, and slope, in its typical form on an x-y graph, represents how two variables relate to each other: rise over run, the change in money over the change in time, etc. In this particular case, the slope we care about describes the relationship between the network’s error and a single weight; i.e. that is, how does the error vary as the weight is adjusted.

From chain rule in calculus we get,

<img src=https://pathmind.com/images/wiki/backprop_chain_rule.png>

These gradients are used to understand how to update each parameter in order to get the local (global) minima. In the NN, parameters are initialized with random values and then updates using the gradients calculated during backpropagation.

<img src=https://miro.medium.com/max/1201/1*VymEfQTf30evUczsTBiz4g.png width=400>

Different optimization techniques are available in the literature. To name a few -

* SGD (stochastic gradient descent)
* Adam
* RMSProp
* Adagrad
* Radam
* Adadelta


### Regularization

Regularization is a very useful technique which is used to generalize the performance of neural networks. To reduce overfitting, we need to restrict the weights of the neurons. There are different regularization parameters - L1, L2 are used to restrict the search space of paremeters.

Another idea is <b>dropout</b>. Using dropout, neural network drop randomly selected neurons during training process. 

<img src="https://miro.medium.com/max/1400/1*iWQzxhVlvadk6VAJjsgXgg.png" width="50%">

Now that we have learn the basics of neural networks, let us build a simple ANN (not from scratch though) using scikit learn. Our objective is to classify the processed clinical notes into different specialities.

In [1]:
import pandas as pd
import numpy as np

pd.options.display.max_colwidth = -1

In [2]:
df = pd.read_csv('clinical_notes_cleaned_pos.csv')

In [3]:
df.head(5)

text  \
0  \n \n \n \n \n <B>CC:</B> Difficulty with word finding.\n \n <B>HX: </B>This 27y/o RHF experienced sudden onset word finding difficulty and slurred speech on the evening of 2/19/96. She denied any associated dysphagia, diplopia, numbness or weakness of her extremities. She went to sleep with her symptoms on 2/19/96, and awoke with them on 2/20/96. She also awoke with a headache (HA) and mild neck stiffness. She took a shower and her HA and neck stiffness resolved. Throughout the day she continued to have difficulty with word finding and had worsening of her slurred speech. That evening, she began to experience numbness and weakness in the lower right face. She felt like there was a &quot;rubber-band&quot; wrapped around her tongue.\n \n For 3 weeks prior to presentation, she experienced transient episodes of a &quot;boomerang&quot; shaped field cut in the left eye. The episodes were not associated with any other symptoms. One week prior to presentation, she went to a local ER for menorrhagia. She had just resumed taking oral birth control pills one week prior to the ER visit after having stopped their use for several months. Local evaluation included an unremarkable carotid duplex scan. However, a HCT with and without contrast reportedly revealed a left frontal gyriform enhancing lesion. An MRI brain scan on 2/20/96 revealed nonspecific white matter changes in the right periventricular region. EEG reportedly showed diffuse slowing. CRP was reportedly &quot;too high&quot; to calibrate.\n \n <B>MEDS:</B> Ortho-Novum 7-7-7 (started 2/3/96), and ASA (started 2/20/96).\n \n <B>PMH:</B> 1)ventral hernia repair 10 years ago, 2)mild &quot;concussion&quot; suffered during a MVA; without loss of consciousness, 5/93, 3) Anxiety disorder, 4) One childbirth.\n \n <B>FHX: </B>She did not know her father and was not in contact with her mother.\n \n <B>SHX:</B> Lives with boyfriend. Smokes one pack of cigarettes every three days and has done so for 10 years. Consumes 6 bottles of beers, one day a week. Unemployed and formerly worked at an herbicide plant.\n \n <B>EXAM: </B>BP150/79, HR77, RR22, 37.4C.\n \n MS: A&O to person, place and time. Speech was dysarthric with mild decreased fluency marked by occasional phonemic paraphasic errors. Comprehension, naming and reading were intact. She was able to repeat, though her repetition was occasionally marked by phonemic paraphasic errors. She had no difficulty with calculation.\n \n CN: VFFTC, Pupils 5/5 decreasing to 3/3. EOM intact. No papilledema or hemorrhages seen on fundoscopy. No RAPD or INO. There was right lower facial weakness. Facial sensation was intact, bilaterally. The rest of the CN exam was unremarkable.\n \n MOTOR: 5/5 strength throughout with normal muscle bulk and tone.\n \n Sensory: No deficits.\n \n Coord/Station/Gait: unremarkable.\n \n Reflexes 2/2 throughout. Plantar responses were flexor, bilaterally.\n \n Gen Exam: unremarkable.\n \n <B>COURSE:</B> CRP 1.2 (elevated), ESR 10, RF 20, ANA 1:40, ANCA &lt;1:40, TSH 2.0, FT4 1.73, Anticardiolipin antibody IgM 10.8GPL units (normal &lt;10.9), Anticardiolipin antibody IgG 14.8GPL (normal&lt;22.9), SSA and SSB were normal. Urine beta-hCG pregnancy and drug screen were negative. EKG, CXR and UA were negative.\n \n MRI brain, 2/21/96 revealed increased signal on T2 imaging in the periventricular white matter region of the right hemisphere. In addition, there were subtle T2 signal changes in the right frontal, right parietal, and left parietal regions as seen previously on her local MRI can. In addition, special FLAIR imaging showed increased signal in the right frontal region consistent with ischemia.\n \n She underwent Cerebral Angiography on 2/22/96. This revealed decreased flow and vessel narrowing the candelabra branches of the RMCA supplying the right frontal lobe. These changes corresponded to the areas of ischemic changes seen on MRI. There was also segmental narrowing of the caliber of the vessels in the circle of

In [4]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vector = TfidfVectorizer(lowercase=True, #this will convert all the tokens into lower case
                         stop_words='english', #remove english stopwords from vocabulary. if we need the stopwords this value should be None
                         analyzer='word', #tokens should be words. we can also use char for character tokens
                         max_features=50000, #maximum vocabulary size to restrict too many features
                         min_df = 5,
                         max_df = .6
                        )

tfidf_vectorized_corpus = tfidf_vector.fit_transform(df.clean_text)

In [5]:
print (tfidf_vectorized_corpus.shape)

(818, 3842)


In [6]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_score, StratifiedKFold

In [7]:
model = MLPClassifier(hidden_layer_sizes=(500,200),activation='relu',solver='adam',learning_rate='adaptive',max_iter=500,learning_rate_init=.001,validation_fraction=0)
cv_scores = cross_val_score(X=tfidf_vectorized_corpus,y=df.label,cv=StratifiedKFold(5,random_state=42),estimator=model,error_score='f1_macro')
print (cv_scores, np.mean(cv_scores))

[0.69879518 0.78181818 0.7607362  0.71604938 0.72222222] 0.7359242327596728


Our basic ANN achieved average F1 score of 74% on the cross validation, which is little worse than logistic and naive bayes. However, ANN performs better than random forest. Tree based methods are widely successful for tabular dataset, however, they fail to capture semantic information from text data.

### References

1. https://towardsdatascience.com/common-loss-functions-in-machine-learning-46af0ffc4d23

2. https://towardsdatascience.com/introducing-deep-learning-and-neural-networks-deep-learning-for-rookies-1-bd68f9cf5883

3. https://pathmind.com/wiki/neural-network#forward

4. https://towardsdatascience.com/complete-guide-of-activation-functions-34076e95d044

5. https://towardsdatascience.com/understanding-neural-networks-19020b758230

6. https://medium.com/@amarbudhiraja/https-medium-com-amarbudhiraja-learning-less-to-learn-better-dropout-in-deep-machine-learning-74334da4bfc5

